File: day6/solution2.py
from collections import Counter

def load_coordinates():
    coords = {}
    with open("input.txt") as fh:
        for i, line in enumerate(fh):
            line = line.rstrip('\n')
            x, y = line.split(', ')
            x, y = int(x), int(y)
            coords[i] = x, y
    return coords

def get_max_coords(coords):
    grid = []
    max_x, max_y = 0, 0
    for x, y in coords.values():
        if x > max_x:
            max_x = x
        if y > max_y:
            max_y = y

    # We need to add 1 to the max x and y to account for the origin
    max_x += 1
    max_y += 1

    return max_x, max_y

def get_safe_region_area(max_x, max_y, coords):
    safe_region = 0
    for row in range(max_y):
        for col in range(max_x):
            dist = sum([ get_manhattan_dist((col, row), coords[point]) for point in coords ])
            if dist < 10000:
                safe_region += 1
    return safe_region

def get_manhattan_dist(start, end):
    output = abs(start[0] - end[0]) + abs(start[1] - end[1])
    return output

def main():
    coords = load_coordinates()
    max_x, max_y = get_max_coords(coords)
    safe_region = get_safe_region_area(max_x, max_y, coords)
    print("Size of the safe region: {}".format(safe_region))

if __name__ == '__main__':
    main()

File: day6/solution2.rs
use std::fs;
use std::io;
use std::io::BufRead;

#[derive(Debug)]
#[derive(Hash)]
#[derive(Copy, Clone)]
struct Point {
    x: i32,
    y: i32,
}

impl PartialEq for Point {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y
    }
}

impl Eq for Point { }

fn get_max_coords(coords: &Vec<Point>) -> (i32, i32) {
    let mut max_x = coords.iter().map(|p| p.x).max().unwrap();
    let mut max_y = coords.iter().map(|p| p.y).max().unwrap();
    max_x += 1;
    max_y += 1;
    (max_x, max_y)
}

fn get_safe_region_area(max_x: i32, max_y: i32, coords: &Vec<Point>) -> i32 {
    let mut safe_region = 0;
    for row in 0..max_x {
        for col in 0..max_y {
            let dist: i32 = coords.iter()
                .map(|p| get_manhattan_dist(&Point { x: col, y: row }, p)).sum();
            if dist < 10000 {
                safe_region += 1;
            }
        }
    }
    safe_region
}

fn get_manhattan_dist(start: &Point, end: &Point) -> i32 {
    let output = i32::abs(start.x - end.x) + i32::abs(start.y - end.y);
    output
}

fn load_coordinates() -> Vec<Point> {
    let mut output = Vec::new();

    let input = fs::File::open("input.txt")
        .expect("Something went wrong reading the file");
    let reader = io::BufReader::new(input);

    for line in reader.lines() {
        let line = line.unwrap().replace(" ", "");
        let coord_vec: Vec<&str> = line.split(',').collect();
        let x: i32 = coord_vec[0].parse().unwrap();
        let y: i32 = coord_vec[1].parse().unwrap();
        let p = Point { x: x, y: y };
        output.push(p);
    }
    output
}

fn main() {
    let coords = load_coordinates();
    let (max_x, max_y) = get_max_coords(&coords);
    let safe_region = get_safe_region_area(max_x, max_y, &coords);
    println!("Size of the safe region: {}", safe_region);
}

